Libraries/Opc.Ua.Client/Subscription.cs

```diff
@@ -1664,10 +1767,33 @@ private void AdjustCounts(ref uint keepAliveCount, ref uint lifetimeCount)
         /// <summary>
         /// Processes the incoming messages.
         /// </summary>
-        private void OnMessageReceived(object state)
+        private async Task OnMessageReceived()
         {
+            SemaphoreSlim semaphore; //Avoid semaphore being replaced for this instance while running, retain reference locally.
+            lock (m_cache)
+            {
+                //Semaphore is maintained under m_cache lock, avoid semaphore swap issues when possible. The wait call will still sync the message workers properly.
+                semaphore = m_messageWorkersSemaphore;
+            } 
+            
+            var needSemaphore = semaphore != null; //Later used to know if releasing the semaphore is needed. Assumed entered if needed.
+            if (needSemaphore)
+            {
+                try
+                {
+                    await semaphore.WaitAsync();
```

Comment:
I wonder if Tasks which are waiting here are really released in order, have you checked that itis guaranteed?

Reply:
The order of the tasks being released is irrelevant. The worker tasks are fungible, doesn't matter which one is released. When sequential processing is enabled, only one will be active at a given time. 

The mechanism for ensuring order is how they pull items off the m_incomingMessages linked list, which gets "ordered" when new messages are inserted into it by SaveMessageInCache. 

Order of threads doesn't matter here.

Reply:
ok, got it, this is just the worker which pulls the next item...